### PR TITLE
dialects: Update support for returning multiple results using stencil.return

### PR DESCRIPTION
**Pull request by meshtag**
This PR intends to update the support for multiple stencil results. The [earlier version](https://github.com/xdslproject/xdsl/pull/648) produced some bugs when lowered via Psyclone, this patch should not face a similar problem. Here's a [link](https://xdsl.zulipchat.com/#narrow/stream/362011-xDSL/topic/stencil.20dialect/near/346783038) to @mesham's original issue related to this. 

edit @georgebisbas : Also check: https://github.com/xdslproject/xdsl/pull/659#issuecomment-1509276662

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
Can we add test cases her? To actually demonstrate that @mesham's testcase works?

**Comment by meshtag** [comment]
> Can we add test cases her? To actually demonstrate that @mesham's testcase works?

Sure I can do this. I just want to know his earlier test case which broke this in the first place.

**Comment by tobiasgrosser** [comment]
> > Can we add test cases her? To actually demonstrate that @mesham's testcase works?
> 
> Sure, I can do this. I just want to know his earlier test case which broke this in the first place.

He might be busy. Maybe you can derive it from his comments (if he does not reply) without too much trouble?

**Comment by meshtag** [comment]
> He might be busy. Maybe you can derive it from his comments (if he does not reply) without too much trouble?

I am not able to easily reproduce it (atleast from the Python land).

**Review by georgebisbas — CHANGES_REQUESTED** [review]
I would keep on hold until we have a test for this

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 1]
```diff
@@ -393,18 +398,20 @@ def map_returns(op: Operation) -> None:
...
+            cast = store.field.owner
 
-        assert isinstance(cast, CastOp)
+            assert isinstance(cast, CastOp)
 
-        return_targets[op] = cast
+            if return_targets.get(op):
+                return_targets[op].append(cast)
+            else:
+                return_targets[op] = [cast]
```
Comment (PapyChacal) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:414]: I think initializing `return_targets[op]` to `[]` before looping on results would be more readable. What do you think?

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 2]
```diff
@@ -393,18 +398,20 @@ def map_returns(op: Operation) -> None:
...
-
-        if (len(res.uses) > 1) or (not isinstance(
-            (store := list(res.uses)[0].operation), StoreOp)):
-            warn("Only single store result atm")
-            return
+        for res in list(apply.res):
+            if (len(res.uses) > 1) or (not isinstance(
+                (store := list(res.uses)[0].operation), StoreOp)):
+                warn("Only single store for a single return op result atm")
+                return
```
Comment (PapyChacal) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:405]: I think we want to get rid of this warning if we support multiple returns. We should just check that they all map to `StoreOp`s still, before we have inlining / automatic buffering passes.

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 3]
```diff
@@ -117,28 +117,32 @@ def collectBlockArguments(number: int, block: Block):
...
 
-        args = collectBlockArguments(dims, block)
+            args = collectBlockArguments(dims, block)
 
-        store = memref.Store.get(op.arg, subview.result, args)
+            store_list.append(memref.Store.get(op.arg[j], subview.result,
+                                               args))
 
-        rewriter.replace_matched_op([store])
+        rewriter.replace_matched_op([*store_list])
```
Comment (PapyChacal) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:145]:
```suggestion
        rewriter.replace_matched_op(store_list)
```

File: tests/filecheck/dialects/stencil/hdiff.mlir [hunk 4]
```diff
@@ -19,79 +20,83 @@
       %cst = "arith.constant"() {"value" = -4.0 : f64} : () -> f64
       %18 = "arith.mulf"(%14, %cst) : (f64, f64) -> f64
       %19 = "arith.addf"(%18, %17) : (f64, f64) -> f64
-      "stencil.return"(%19) : (!stencil.result<f64>) -> ()
-    }) : (!stencil.temp<[-1 : i64, -1 : i64, -1 : i64], f64>) -> !stencil.temp<[-1 : i64, -1 : i64, -1 : i64], f64>
-    "stencil.store"(%8, %4) {"lb" = #stencil.index<[0 : i64, 0 : i64, 0: i64]>, "ub" = #stencil.index<[64 : i64, 64 : i64, 64 : i64]>} : (!stencil.temp<[-1 : i64, -1 : i64, -1 : i64], f64>, !stencil.field<[72 : i64, 72 : i64, 72 : i64], f64>) -> ()
+      "stencil.return"(%19, %18) : (!stencil.result<f64>, !stencil.result<f64>) -> ()
```
Comment (PapyChacal) [inline tests/filecheck/dialects/stencil/hdiff.mlir:23]: As we're not using `stencil.store_result` here:
```suggestion
      "stencil.return"(%19, %18) : (f64, f64) -> ()
```

**Review by PapyChacal — COMMENTED** [review]
Some minor comments, but looks sensible! Thanks!

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py (hunk 2 above)
Comment (meshtag) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:405]: For the case of inlining, it is not necessary that all results of `stencil.apply` will map to `StoreOp`. 
For ex. 
```
...

%1, %2 = stencil.apply(...) {...}

%3, %4 = stencil.apply(%1, ...) {...}

stencil.store(%1, %0) {...}

...
```

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py (hunk 3 above)
Comment (meshtag) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:145]: I get this pylance type issue if I change it like that 
```
Argument of type "list[Store]" cannot be assigned to parameter "new_ops" of type "Operation | list[Operation]" in function "replace_matched_op"
  Type "list[Store]" cannot be assigned to type "Operation | list[Operation]"
    "list[Store]" is incompatible with "Operation"
    TypeVar "_T@list" is invariant
      "Store" is incompatible with "Operation"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)
```

Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:145]: Oh, I see. Let's stick with that for now then.

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:406]: Right, so we could get rid of this whole check and warning?

**Review by PapyChacal — APPROVED** [review]

**Comment by PapyChacal** [comment]
We might just want to add a test for `stencil.ReturnOp.get` now!
You can have a look at similar tests in `tests/dialects` if you feel like adding it, but I'm happy to add it tomorrow otherwise

**Comment by georgebisbas** [comment]
> This should ideally fix the problem related to `stencil.return` op build.
Could you point to this problem?

**Comment by meshtag** [comment]
> Could you point to this problem?

There was some issue faced by @mesham. @mesham, can you please check if the problem still persists. Please state how to reproduce the issue as well (in case it is unsolved).

**Comment by PapyChacal** [comment]
If I'm not mistaken, @mesham is having time off now, so I'll answer my understanding on that:

@mesham had example code where having multiple returns would make sense, and tried to generate and lower them, as the API suggested it was possible. This was not implemented yet though (especially the lowering part) and the `get` was erroring, which led to think that `get` was the issue.

Meanwhile, he generated his code by splitting all stencils with multiple outputs in seperate stencils and got it working, so this PR shouldn't have any regression on that side. I just tested on the latest IR sample I got from Psyclone and it lowered just fine :)

**Comment by PapyChacal** [comment]
@georgebisbas here's a link to @mesham's original issue : https://xdsl.zulipchat.com/#narrow/stream/362011-xDSL/topic/stencil.20dialect/near/346783038. Your request for changes is blocking the merge, is it on purpose?

**Comment by georgebisbas** [comment]
Thanks, ideally in these, add some description to this PR, so that someone understands what happens.
In the most ideal world, we would need an MFE of NIck's issue.
But at least two lines of description should be enough.

**Comment by meshtag** [comment]
> This was not implemented yet though (especially the lowering part) and the get was erroring, which led to think that get was the issue.

Can you please specify what is referred to as "lowering part" here? 

> Thanks, ideally in these, add some description to this PR, so that someone understands what happens.

Sure! I was not sure about the bug details and how to reproduce it at the time of this PR's creation (am still not very sure tbh, will trust @PapyChacal ). It was sort of an ad-hoc solution to see if we could solve the problem. Nevertheless, the original patch's merge was reverted and I am now changing this PR's purpose to "adding the multiple return support again". I updated the description to mention this. Let me know if I should change anything else.

**Comment by PapyChacal** [comment]
By "the lowering part", I mean lowering a stencil.return with multiple operands, to be able to lower a stencil.apply outputting multiple stencil.temp

**Comment by PapyChacal** [comment]
But, to sum it up, this PR:

- Generalize the stencil.return lowering to handle multiple in a backward-compatible manner (all single-operands returns are lowered just as before)
- Change the stencil.ReturnOp operand to be variadic(as it is intended to), so this might be the slightly breaking change, as it now should be passed a length 1 list for a single return. It also introduces a test, where it was untested before. 

Writing this I realize the new constructor could be made to work with single values too, so the PR is completely backward compatible.

**Comment by georgebisbas** [comment]
> But, to sum it up, this PR:
> 
> * Generalize the stencil.return lowering to handle multiple in a backward-compatible manner (all single-operands returns are lowered just as before)
> * Change the stencil.ReturnOp operand to be variadic(as it is intended to), so this might be the slightly breaking change, as it now should be passed a length 1 list for a single return. It also introduces a test, where it was untested before.
> 
> Writing this I realize the new constructor could be made to work with single values too, so the PR is completely backward compatible.

ok, added a link to this comment in the first post

**Review by georgebisbas — APPROVED** [review]

**Comment by meshtag** [comment]
Can we merge this?

**Comment by PapyChacal** [comment]
> Can we merge this?

Yes, I'm on it :slightly_smiling_face:

**Comment by meshtag** [comment]
I'm curious, is it a necessary/advisable requirement to merge main before letting this (or any other PR for that matter) in? I was under the impression that we are fine unless we observe merge conflicts, git should automatically take care of merging things properly unless there is a merge conflict?

**Pull request merged**